README.md
# Extensions Library for .NET

VivaldiSoft.Extensions is a cross-platform open-source library which provides extensions methods to .NET developers.

[![Build Status](https://travis-ci.org/VivaldiSoft/VivaldiSoft.Extensions.svg?branch=master)](https://travis-ci.org/VivaldiSoft/VivaldiSoft.Extensions)
[![NuGet](https://img.shields.io/nuget/v/VivaldiSoft.Extensions.svg)](https://www.nuget.org/packages/VivaldiSoft.Extensions/)
[![BCH compliance](https://bettercodehub.com/edge/badge/VivaldiSoft/VivaldiSoft.Extensions?branch=master)](https://bettercodehub.com/)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

## Installation

VivaldiSoft.Extensions runs on Windows, Linux, and macOS.

Once you have an app, you can install the VivaldiSoft.Extensions NuGet package from the NuGet package manager:

```
Install-Package VivaldiSoft.Extensions
```

Or alternatively you can add the VivaldiSoft.Extensions package from within Visual Studio's NuGet package manager.

## Examples

Please find examples and the documentation at the [wiki](https://github.com/VivaldiSoft/VivaldiSoft.Extensions/wiki) of this repository.

## Contributing

We welcome contributions! Please review our [contribution guide](CONTRIBUTING.md).

## License

VivaldiSoft.Extensions is licensed under the [MIT license](LICENSE).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2019-03-26
### Added
- ArrayExtensions, Base64Extensions, ByteExtensions, DataTableExtensions, DateTimeExtensions, IntegerExtensions, ListExtensions and StringExtensions.


tests/VivaldiSoft.Extensions.Tests/Models/ByteExtensionsTest.cs
using VivaldiSoft.Extensions;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;
using Xunit;

namespace VivaldiSoft.Extensions.Tests
{
    public class ByteExtensionsTest
    {
        [Theory]
        [ClassData(typeof(BytesToHexaData))]
        public void BytesToHexa(byte[] value, string separator, string expected)
        {
            var result = value.BytesToHexa(separator);

            Assert.Equal(expected, result);
        }

        [Theory]
        [ClassData(typeof(HexaToBytesData))]
        public void HexaToBytes(string value, string separator, byte[] expected)
        {
            var result = value.HexaToBytes(separator);

            Assert.Equal(expected, result);
        }

        [Theory]
        [ClassData(typeof(BytesToStringData))]
        public void BytesToString(byte[] value, string expected)
        {
            var result = value.BytesToString();

            Assert.Equal(expected, result);
        }

        [Fact]
        public void BytesToString_Exception_value_Null()
        {
            byte[] value = null;

            Assert.Throws<ArgumentNullException>(() => ByteExtensions.BytesToString(value));
        }

        [Theory]
        [ClassData(typeof(StringToBytesData))]
        public void StringToBytes(string value, byte[] expected)
        {
            var result = value.StringToBytes();

            Assert.Equal(expected, result);
        }

        [Fact]
        public void StringToBytes_Exception_value_Null()
        {
            string value = null;

            Assert.Throws<ArgumentNullException>(() => ByteExtensions.StringToBytes(value));
        }

        public class BytesToHexaData : IEnumerable<object[]>
        {
            public IEnumerator<object[]> GetEnumerator()
            {
                yield return new object[] { new byte[] { }, "", "" };
                yield return new object[] { new byte[] { 48, 49, 50, 51, 52 }, "", "3031323334" };
                yield return new object[] { new byte[] { 48, 49, 50, 51, 52 }, "-", "30-31-32-33-34" };
            }

            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
        }

        public class HexaToBytesData : IEnumerable<object[]>
        {
            public IEnumerator<object[]> GetEnumerator()
            {
                yield return new object[] { "", "", new byte[] { } };
                yield return new object[] { "3031323334", "", new byte[] { 48, 49, 50, 51, 52 } };
                yield return new object[] { "30-31-32-33-34", "-", new byte[] { 48, 49, 50, 51, 52 } };
            }

            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
        }

        public class BytesToStringData : IEnumerable<object[]>
        {
            public IEnumerator<object[]> GetEnumerator()
            {
                yield return new object[] { new byte[] { }, "" };
                yield return new object[] { new byte[] { 48, 49, 50, 51, 52 }, "01234" };
                yield return new object[] { new byte[] { 97, 98, 99, 100, 101 }, "abcde" };
            }

            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
        }

        public class StringToBytesData : IEnumerable<object[]>
        {
            public IEnumerator<object[]> GetEnumerator()
            {
                yield return new object[] { "", new byte[] { } };
                yield return new object[] { "01234", new byte[] { 48, 49, 50, 51, 52 } };
                yield return new object[] { "abcde", new byte[] { 97, 98, 99, 100, 101 } };
            }

            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
        }
    }
}


tests/VivaldiSoft.Extensions.Tests/Models/ListExtensionsTest.cs
using VivaldiSoft.Extensions;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Xunit;

namespace VivaldiSoft.Extensions.Tests
{
    public class ListExtensionsTest
    {
        [Theory]
        [ClassData(typeof(IndexOfMaxData))]
        public void IndexOfMax(double[] values, int expected)
        {
            var result = values.IndexOfMax();

            Assert.Equal(expected, result);
        }

        [Theory]
        [ClassData(typeof(IndexOfMinData))]
        public void IndexOfMin(double[] values, int expected)
        {
            var result = values.IndexOfMin();

            Assert.Equal(expected, result);
        }

        [Theory]
        [ClassData(typeof(IsSameData))]
        public void IsSame(double[] first, double[] second, bool expected)
        {
            var result = first.IsSame(second);

            Assert.Equal(expected, result);
        }

        [Theory]
        [ClassData(typeof(ShuffleData))]
        public void Shuffle(double[] values)
        {
            double[] original = new double[values.Length];
            Array.Copy(values, 0, original, 0, values.Length);

            values.Shuffle();

            // check length
            bool ok = values.Length == original.Length;

            if (ok)
            {
                ok = false;

                if (values.Length == 1)
                {
                    ok = true;
                }
                else
                {
                    // check shuffle
                    for (int i = 0; i < values.Length; i++)
                    {
                        if (values[i] != original[i])
                        {
                            ok = true;
                            break;
                        }
                    }
                }

                // check same elements
                for (int i = 0; i < values.Length; i++)
                {
                    int countValues = 0;
                    int countOriginal = 0;

                    for (int j = 0; j < values.Length; j++)
                    {
                        if (values[j] == values[i])
                        {
                            countValues++;
                        }
                    }

                    for (int j = 0; j < original.Length; j++)
                    {
                        if (original[j] == values[i])
                        {
                            countOriginal++;
                        }
                    }

                    if (countValues != countOriginal)
                    {
                        ok = false;
                        break;
                    }
                }
            }

            Assert.True(ok);
        }

        [Theory]
        [ClassData(typeof(MedianData))]
        public void Median(double[] values, double expected)
        {
            var result = values.Median();

            Assert.Equal(expected, result);
        }

        [Fact]
        public void Median_Exception_Value_Null()
        {
            Assert.Throws<ArgumentNullException>(() => ListExtensions.Median(null));
        }

        [Theory]
        [ClassData(typeof(MeanData))]
        public void Mean(double[] values, double expected)
        {
            List<double> list = new List<double>();
            list.AddRange(values);
            var result = list.Mean();

            Assert.Equal(expected, result);
        }

        [Theory]
        [ClassData(typeof(MeanDelimitedData))]
        public void MeanDelimited(double[] values, int start, int end, double expected)
        {
            List<double> list = new List<double>();
            list.AddRange(values);
            var result = list.Mean(start, end);

            Assert.Equal(expected, result);
        }

        [Fact]
        public void MeanDelimited_Exception_Start_OutOfRange()
        {
            List<double> list = new List<double>() { 0, 1 };
            Assert.Throws<ArgumentOutOfRangeException>(() => list.Mean(1, 0));
        }

        [Fact]
        public void MeanDelimited_Exception_Start_OutOfRange_Bottom()
        {
            List<double> list = new List<double>() { 0, 1 };
            Assert.Throws<ArgumentOutOfRangeException>(() => list.Mean(-1, 1));
        }

        [Fact]
        public void MeanDelimited_Exception_Start_OutOfRange_Up()
        {
            List<double> list = new List<double>() { 0, 1 };
            Assert.Throws<ArgumentOutOfRangeException>(() => list.Mean(2, 1));
        }

        [Fact]
        public void MeanDelimited_Exception_End_OutOfRange_Bottom()
        {
            List<double> list = new List<double>() { 0, 1 };
            Assert.Throws<ArgumentOutOfRangeException>(() => list.Mean(0, -1));
        }

        [Fact]
        public void MeanDelimited_Exception_End_OutOfRange_Up()
        {
            List<double> list = new List<double>() { 0, 1 };
            Assert.Throws<ArgumentOutOfRangeException>(() => list.Mean(0, 2));
        }

        public class IndexOfMaxData : IEnumerable<object[]>
        {
            public IEnumerator<object[]> GetEnumerator()
            {
                yield return new object[] { new double[] { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 }, 9 };
                yield return new object[] { new double[] { 9, 8, 7, 6, 5, 4, 3, 2, 1, 0 }, 0 };
                yield return new object[] { new double[] { 0, 1, 2, 3, 4, 5, 4, 3, 2, 1, 0 }, 5 };
                yield return new object[] { new double[] { 0, 0, 1, 1, 2, 2, 1, 1, 0, 0 }, 5 };
            }

            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
        }

        public class IndexOfMinData : IEnumerable<object[]>
        {
            public IEnumerator<object[]> GetEnumerator()
            {
                yield return new object[] { new double[] { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 }, 0 };
                yield return new object[] { new double[] { 9, 8, 7, 6, 5, 4, 3, 2, 1, 0 }, 9 };
                yield return new object[] { new double[] { 5, 4, 3, 2, 1, 0, 1, 2, 3, 4, 5 }, 5 };
                yield return new object[] { new double[] { 2, 2, 1, 1, 0, 0, 1, 1, 2, 2 }, 5 };
            }

            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
        }

        public class ShuffleData : IEnumerable<object[]>
        {
            public IEnumerator<object[]> GetEnumerator()
            {
                yield return new object[] { new double[] { 0 } };
                yield return new object[] { new double[] { 0, 0, 1, 1, 2, 2, 3, 3, 4, 4, 5, 5 } };
            }

            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
        }

        public class IsSameData : IEnumerable<object[]>
        {
            public IEnumerator<object[]> GetEnumerator()
            {
                yield return new object[] { new double[] { }, new double[] { }, true };
                yield return new object[] { new double[] { 0 }, new double[] { 0 }, true };
                yield return new object[] { new double[] { 0, 1 }, new double[] { 0, 1 }, true };
                yield return new object[] { new double[] { 0, 1 }, new double[] { 1, 0 }, true };
                yield return new object[] { new double[] { 0 }, new double[] { 0, 1 }, false };
                yield return new object[] { new double[] { 0, 1 }, new double[] { 0 }, false };
            }

            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
        }

        public class MedianData : IEnumerable<object[]>
        {
            public IEnumerator<object[]> GetEnumerator()
            {
                yield return new object[] { new double[] { 0 }, 0 };
                yield return new object[] { new double[] { 1 }, 1 };
                yield return new object[] { new double[] { 2, 3, 5, 1, 4 }, 3 };
                yield return new object[] { new double[] { 2, 1, 4, 5, 3, 0 }, 2.5 };
            }

            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
        }

        public class MeanData : IEnumerable<object[]>
        {
            public IEnumerator<object[]> GetEnumerator()
            {
                yield return new object[] { new double[] { }, 0 };
                yield return new object[] { new double[] { 0 }, 0 };
                yield return new object[] { new double[] { 1 }, 1 };
                yield return new object[] { new double[] { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 }, 4.5 };
            }

            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
        }

        public class MeanDelimitedData : IEnumerable<object[]>
        {
            public IEnumerator<object[]> GetEnumerator()
            {
                yield return new object[] { new double[] { 0 }, 0, 0, 0 };
                yield return new object[] { new double[] { 1 }, 0, 0, 1 };
                yield return new object[] { new double[] { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 }, 0, 9, 4.5 };
                yield return new object[] { new double[] { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 }, 1, 8, 4.5 };
                yield return new object[] { new double[] { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 }, 0, 5, 2.5 };
            }

            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
        }
    }
}


tests/VivaldiSoft.Extensions.Tests/Models/ArrayExtensionsTest.cs
using VivaldiSoft.Extensions;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;
using Xunit;

namespace VivaldiSoft.Extensions.Tests
{
    public class ArrayExtensionsTest
    {
        [Theory]
        [ClassData(typeof(ArrayEqualData))]
        public void ArrayEqual(double[] first, double[] second, bool expected)
        {
            var result = first.ArrayEqual(second);

            Assert.Equal(expected, result);
        }

        [Theory]
        [ClassData(typeof(CombineData))]
        public void Combine(double[] first, double[] second, double[] expected)
        {
            var result = first.Combine(second);

            Assert.Equal(expected, result);
        }

        [Theory]
        [ClassData(typeof(SetAllValuesData))]
        public void SetAllValues(double[] value, double setter, double[] expected)
        {
            var result = value.SetAllValues(setter);

            Assert.Equal(expected, result);
        }

        [Theory]
        [ClassData(typeof(SplitByNumberOfDivisionsData))]
        public void SplitByNumberOfDivisions(double[] value, int numberOfDivisions, double[][] expected)
        {
            var result = value.SplitByNumberOfDivisions(numberOfDivisions);

            Assert.Equal(expected, result);
        }

        [Fact]
        public void SplitByNumberOfDivisions_Exception_ZeroDivisions()
        {
            Assert.Throws<DivideByZeroException>(() => ArrayExtensions.SplitByNumberOfDivisions(new double[0], 0));
        }

        [Theory]
        [ClassData(typeof(SplitByNumberOfElementsInDivisionData))]
        public void SplitByNumberOfElementsInDivision(double[] value, int elementsInDivision, double[][] expected)
        {
            var result = value.SplitByNumberOfElementsInDivision(elementsInDivision);

            Assert.Equal(expected, result);
        }

        [Fact]
        public void SplitByNumberOfElementsInDivision_Exception_ZeroDivisions()
        {
            Assert.Throws<DivideByZeroException>(() => ArrayExtensions.SplitByNumberOfElementsInDivision(new double[0], 0));
        }

        [Theory]
        [ClassData(typeof(SubArrayData))]
        public void SubArray(double[] value, int index, int length, double[] expected)
        {
            var result = value.SubArray(index, length);

            Assert.Equal(expected, result);
        }

        [Fact]
        public void SubArray_Exception_value_Null()
        {
            double[] list = null;

            Assert.Throws<ArgumentNullException>(() => ArrayExtensions.SubArray(list, 0, 1));
        }

        [Fact]
        public void SubArray_Exception_Array_OutOfRange_Bottom()
        {
            double[] list = new double[] { 0, 1 };
            Assert.Throws<ArgumentOutOfRangeException>(() => ArrayExtensions.SubArray(list, -1, 1));
        }

        public class ArrayEqualData : IEnumerable<object[]>
        {
            public IEnumerator<object[]> GetEnumerator()
            {
                yield return new object[] { new double[] { 0 }, new double[] { 0 }, true };
                yield return new object[] { new double[] { 0, 1 }, new double[] { 0, 1 }, true };
                yield return new object[] { new double[] { 0 }, new double[] { 0, 1 }, false };
                yield return new object[] { new double[] { }, new double[] { }, true };
                yield return new object[] { new double[] { }, new double[] { 0 }, false };
                yield return new object[] { new double[] { 0 }, new double[] { }, false };
            }

            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
        }

        public class CombineData : IEnumerable<object[]>
        {
            public IEnumerator<object[]> GetEnumerator()
            {
                yield return new object[] { new double[] { 0, 1, 2 }, new double[] { 3, 4, 5 }, new double[] { 0, 1, 2, 3, 4, 5 } };
                yield return new object[] { new double[] { 0 }, new double[] { }, new double[] { 0 } };
                yield return new object[] { new double[] { }, new double[] { 0 }, new double[] { 0 } };
            }

            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
        }

        public class SetAllValuesData : IEnumerable<object[]>
        {
            public IEnumerator<object[]> GetEnumerator()
            {
                yield return new object[] { new double[] { }, 1, new double[] { } };
                yield return new object[] { new double[] { 0 }, 1, new double[] { 1 } };
                yield return new object[] { new double[] { 0, 0, 0, 0, 0 }, 1, new double[] { 1, 1, 1, 1, 1 } };
                yield return new object[] { new double[] { 0, 0, 0, 0, 0 }, 1.2, new double[] { 1.2, 1.2, 1.2, 1.2, 1.2 } };
            }

            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
        }

        public class SplitByNumberOfDivisionsData : IEnumerable<object[]>
        {
            public IEnumerator<object[]> GetEnumerator()
            {
                yield return new object[] { new double[] { }, 1,
                    new double[0][]
                };
                yield return new object[] { new double[] { 0 }, 1,
                    new double[][] {
                        new double[] { 0 }
                    }
                };
                yield return new object[] { new double[] { 0 }, 2,
                    new double[][] {
                        new double[] { 0 }
                    }
                };
                yield return new object[] { new double[] { 0, 1 }, 1,
                    new double[][] {
                        new double[] { 0, 1 }
                    }
                };
                yield return new object[] { new double[] { 0, 1 }, 2,
                    new double[][] {
                        new double[] { 0 },
                        new double[] { 1 }
                    }
                };
                yield return new object[] { new double[] { 0, 1, 2 }, 2,
                    new double[][] {
                        new double[] { 0, 1 },
                        new double[] { 2 }
                    }
                };
            }

            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
        }

        public class SplitByNumberOfElementsInDivisionData : IEnumerable<object[]>
        {
            public IEnumerator<object[]> GetEnumerator()
            {
                yield return new object[] { new double[] { }, 1,
                    new double[0][]
                };
                yield return new object[] { new double[] { 0 }, 1,
                    new double[][] {
                        new double[] { 0 }
                    }
                };
                yield return new object[] { new double[] { 0 }, 2,
                    new double[][] {
                        new double[] { 0 }
                    }
                };
                yield return new object[] { new double[] { 0, 1 }, 1,
                    new double[][] {
                        new double[] { 0 },
                        new double[] { 1 }
                    }
                };
                yield return new object[] { new double[] { 0, 1 }, 2,
                    new double[][] {
                        new double[] { 0, 1 }
                    }
                };
                yield return new object[] { new double[] { 0, 1, 2 }, 2,
                    new double[][] {
                        new double[] { 0, 1 },
                        new double[] { 2 }
                    }
                };
            }

            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
        }

        public class SubArrayData : IEnumerable<object[]>
        {
            public IEnumerator<object[]> GetEnumerator()
            {
                yield return new object[] { new double[] { 0, 1, 2, 3, 4, 5 }, 0, 6, new double[] { 0, 1, 2, 3, 4, 5 } };
                yield return new object[] { new double[] { 0, 1, 2, 3, 4, 5 }, 1, 4, new double[] { 1, 2, 3, 4 } };
                yield return new object[] { new double[] { 0, 1, 2, 3, 4, 5 }, 0, 1, new double[] { 0 } };
                yield return new object[] { new double[] { 0, 1, 2, 3, 4, 5 }, 5, 1, new double[] { 5 } };
                yield return new object[] { new double[] { 0, 1, 2, 3, 4, 5 }, 0, 0, new double[] { } };
            }

            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
        }
    }
}


src/VivaldiSoft.Extensions/Models/ArrayExtensions.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace VivaldiSoft.Extensions
{
    /// <summary>
    /// A static class of extension methods for <see cref="Array"/>.
    /// </summary>
    public static class ArrayExtensions
    {
        /// <summary>
        /// Checks if the Arrays are equal.
        /// </summary>
        /// <typeparam name="T">Array type.</typeparam>
        /// <param name="first">The <see cref="Array"/> that contains data to compare with.</param>
        /// <param name="second">The <see cref="Array"/> that contains data to compare to.</param>
        /// <returns>Returns <c>true</c> if all element match and <c>false</c> otherwise.</returns>
        /// <example>
        /// <code lang="csharp">
        /// var first= new double[] { 0, 0, 0, 0, 0 };
        /// var second = new double[] { 0, 0, 0, 0, 0 };
        /// var result = first.ArrayEqual(second);
        /// 
        /// /*
        /// result is true
        /// */
        /// </code>
        /// </example>
        public static bool ArrayEqual<T>(this T[] first, T[] second) where T : IEquatable<T>
        {
            if (ReferenceEquals(first, second))
            {
                return true;
            }

            if (first is null || second is null)
            {
                return false;
            }

            if (first.Length != second.Length)
            {
                return false;
            }

            for (int i = 0; i < first.Length; i++)
            {
                if (!first[i].Equals(second[i]))
                {
                    return false;
                }
            }

            return true;
        }

        /// <summary>
        /// Combine the Arrays.
        /// </summary>
        /// <typeparam name="T">Array type.</typeparam>
        /// <param name="first">First element to combine.</param>
        /// <param name="second">Second element to combine.</param>
        /// <returns>T[].</returns>
        /// <example>
        /// <code lang="csharp">
        /// var first= new double[] { 0, 1, 2 };
        /// var second = new double[] { 3, 4, 5 };
        /// var result = first.Combine(second);
        /// 
        /// /*
        /// result contains the values { 0, 1, 2, 3, 4, 5 }
        /// */
        /// </code>
        /// </example>
        public static T[] Combine<T>(this T[] first, T[] second)
        {
            T[] result = new T[first.Length + second.Length];
            Array.Copy(first, 0, result, 0, first.Length);
            Array.Copy(second, 0, result, first.Length, second.Length);
            return result;
        }

        /// <summary>
        /// Sets all values.
        /// </summary>
        /// <typeparam name="T">The type of the elements of the array that will be modified.</typeparam>
        /// <param name="array">The one-dimensional, zero-based array.</param>
        /// <param name="value">The value.</param>
        /// <returns>A reference to the changed array.</returns>
        /// <example>
        /// <code lang="csharp">
        /// var array = new double[] { 0, 0, 0, 0, 0 };
        /// var result = array.SetAllValues(1);
        /// 
        /// /*
        /// result contains the values { 1, 1, 1, 1, 1 }
        /// */
        /// </code>
        /// </example>
        public static T[] SetAllValues<T>(this T[] array, T value)
        {
            for (int i = 0; i < array.Length; i++)
            {
                array[i] = value;
            }

            return array;
        }

        /// <summary>
        /// Split an array into subarrays by specifying the number of divisions.
        /// </summary>
        /// <typeparam name="T">The type of the elements of the array that will be split.</typeparam>
        /// <param name="array">Array to split.</param>
        /// <param name="numberOfDivisions">Number of divisions.</param>
        /// <returns>Subarrays.</returns>
        /// <exception cref="DivideByZeroException">Number of divisions is zero.</exception>
        /// <example>
        /// <code lang="csharp">
        /// var array = new double[] { 0, 1, 2 };
        /// var numberOfDivisions = 2;
        /// var result = array.SplitByNumberOfDivisions(numberOfDivisions);
        /// 
        /// /*
        /// result contains the values { { 0, 1 }, { 2 } }
        /// */
        /// </code>
        /// </example>
        public static IEnumerable<T[]> SplitByNumberOfDivisions<T>(this T[] array, int numberOfDivisions)
        {
            if (numberOfDivisions < 1)
            {
                throw (new DivideByZeroException("numberDivisions"));
            }

            IEnumerable<T[]> result;
            int elementsInDivision = (int)Math.Ceiling(array.Length / (double)numberOfDivisions);

            if (elementsInDivision > 0)
            {
                result = array.SplitByNumberOfElementsInDivision(elementsInDivision);
            }
            else
            {
                result = new List<T[]>();
            }

            return result;
        }

        /// <summary>
        /// Split an array into subarrays by specifying the number of elements in each division.
        /// </summary>
        /// <typeparam name="T">The type of the elements of the array that will be split.</typeparam>
        /// <param name="array">Array to split.</param>
        /// <param name="elementsInDivision">Elements in each division.</param>
        /// <returns>Subarrays.</returns>
        /// <exception cref="DivideByZeroException">Number of elements in division is zero.</exception>
        /// <example>
        /// <code lang="csharp">
        /// var array = new double[] { 0, 1, 2 };
        /// var elementsInDivision = 2;
        /// var result = array.SplitByNumberOfElementsInDivision(elementsInDivision);
        /// 
        /// /*
        /// result contains the values { { 0, 1 }, { 2 } }
        /// */
        /// </code>
        /// </example>
        public static IEnumerable<T[]> SplitByNumberOfElementsInDivision<T>(this T[] array, int elementsInDivision)
        {
            if (elementsInDivision < 1)
            {
                throw (new DivideByZeroException("elementsInDivision"));
            }

            var result = new List<T[]>();

            for (int index = 0; index < array.Length; index += elementsInDivision)
            {
                var length = index + elementsInDivision < array.Length ? elementsInDivision : array.Length - index;
                var temp = new T[length];
                Array.Copy(array, index, temp, 0, length);
                result.Add(temp);
            }

            return result;
        }

        /// <summary>
        /// Subs the array.
        /// </summary>
        /// <typeparam name="T">The type of the elements of the array that will be modified.</typeparam>
        /// <param name="array">The data.</param>
        /// <param name="index">The index.</param>
        /// <param name="length">The length.</param>
        /// <returns>T[].</returns>
        /// <example>
        /// <code lang="csharp">
        /// var array = new double[] { 0, 1, 2, 3, 4, 5 };
        /// var index = 1;
        /// var length = 4;
        /// var result = array.SubArray(index, length);
        /// 
        /// /*
        /// result contains the values { 1, 2, 3, 4 }
        /// */
        /// </code>
        /// </example>
        public static T[] SubArray<T>(this T[] array, int index, int length)
        {
            T[] result = new T[length];
            Array.Copy(array, index, result, 0, length);
            return result;
        }
    }
}


src/VivaldiSoft.Extensions/Models/ByteExtensions.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace VivaldiSoft.Extensions
{
    /// <summary>
    /// A static class of extension methods for <see cref="Byte"/> <see cref="Array"/>.
    /// </summary>
    public static class ByteExtensions
    {
        /// <summary>
        /// Converts the numeric value of a byte array to its equivalent hexadecimal string.
        /// </summary>
        /// <param name="bytes">The value.</param>
        /// <param name="separator">Hexadecimal values separator.</param>
        /// <returns>Hexadecimal string.</returns>
        /// <example>
        /// <code lang="csharp">
        /// var array = new byte[] { 48, 49, 50, 51, 52 };
        /// var separator = "-";
        /// var result = array.BytesToHexa(separator);
        /// 
        /// /*
        /// result is "30-31-32-33-34"
        /// */
        /// </code>
        /// </example>
        public static string BytesToHexa(this byte[] bytes, string separator = "")
        {
            string sHexa = "";

            for (var i = 0; i < bytes.Length; i++)
            {
                if (!string.IsNullOrEmpty(sHexa))
                {
                    sHexa += separator;
                }
                sHexa += bytes[i].ToString("X").PadLeft(2, '0');
            }

            return sHexa;
        }

        /// <summary>
        /// Converts the hexadecimal string to its quivalent numeric value byte array.
        /// </summary>
        /// <param name="value">The value.</param>
        /// <param name="separator">Hexadecimal values separator.</param>
        /// <returns>Byte array.</returns>
        /// <example>
        /// <code lang="csharp">
        /// var value = "30-31-32-33-34";
        /// var separator = "-";
        /// var result = value.HexaToBytes(separator);
        /// 
        /// /*
        /// result contains the values { 48, 49, 50, 51, 52 }
        /// */
        /// </code>
        /// </example>
        public static byte[] HexaToBytes(this string value, string separator = "")
        {
            List<byte> bytes = new List<byte>();
            if (separator != "")
            {
                value = value.Replace(separator, "");
            }

            for (int i = 0; i < value.Length; i += 2)
            {
                bytes.Add(Convert.ToByte(value.Substring(i, 2), 16));
            }

            return bytes.ToArray();
        }

        /// <summary>
        /// Decodes a sequence of bytes from the specified byte array into a string.
        /// </summary>
        /// <param name="bytes">The byte array containing the sequence of bytes to decode.</param>
        /// <returns>A string that contains the results of decoding the specified sequence of bytes.</returns>
        /// <remarks>Equivalent to System.Text.Encoding.UTF8.GetString().</remarks>
        /// <example>
        /// <code lang="csharp">
        /// var array = new byte[] { 48, 49, 50, 51, 52 };
        /// var result = array.BytesToString();
        /// 
        /// /*
        /// result is "01234"
        /// */
        /// </code>
        /// </example>
        public static string BytesToString(this byte[] bytes)
        {
            return Encoding.UTF8.GetString(bytes);
        }

        /// <summary>
        /// Encodes a set of characters into a sequence of bytes.
        /// </summary>
        /// <param name="value">The character array containing the characters to encode.</param>
        /// <returns>A byte array containing the results of encoding the specified set of characters.</returns>
        /// <remarks>Equivalent to System.Text.Encoding.UTF8.GetBytes().</remarks>
        /// <example>
        /// <code lang="csharp">
        /// var value = "01234";
        /// var result = value.StringToBytes();
        /// 
        /// /*
        /// result contains the values { 48, 49, 50, 51, 52 }
        /// */
        /// </code>
        /// </example>
        public static byte[] StringToBytes(this string value)
        {
            return Encoding.UTF8.GetBytes(value);
        }
    }
}
